module/zfs/dsl_dataset.c

```diff
@@ -3087,7 +3087,11 @@ dsl_dataset_handoff_check(dsl_dataset_t *ds, void *owner, dmu_tx_t *tx)
 		dsl_dataset_long_rele(ds, owner);
 	}
 
-	held = dsl_dataset_long_held(ds);
+	dsl_dir_t *dd = ds->ds_dir;
+	mutex_enter(&dd->dd_activity_lock);
+	if (zfs_refcount_count(&ds->ds_longholds) != dd->dd_activity_count)
```

Comment:
What prevents dd_activity_count from changing just after we drop the dd_activity_lock, invalidating our check?

Note that the hold count can't change because we have the pool config lock for writer (which we should assert and explain here).  Maybe that also prevents the activity_count from changing?

Reply:
That does prevent the activity_count from increasing after we drop the dd_activity_lock, but we're also fine with the dd_activity count changing. Basically this check just verifies that only the waiters have long-holds, and we're going to cancel them before we do anything else, so we don't care if a few more come in or some of them leave.

Reply:
Makes sense.  Could you add a comment explaining that?